File: protokimica-portal/backend/app/core/areas.py
"""
Las áreas de la empresa. Fuente única del backend.

Antes vivían repetidas en seis archivos del frontend y con listas distintas
entre módulos: PQRS conocía "Facturación" y Master Planner no, así que un
indicador de Facturación no cruzaba con nada.

El gemelo de este archivo es `frontend/src/core/areas.js`, y una prueba
verifica que los dos digan exactamente lo mismo.

**La escritura exacta importa.** El área se compara como texto en varios
sitios (por ejemplo, quién puede cerrar una PQRS), así que "Servicio al
Cliente" y "Servicio al cliente" son áreas distintas para el sistema. Si hay
que cambiar cómo se escribe una, va con migración de datos.

Cuando el portal se venda a más de una empresa esto pasa a ser una tabla por
tenant. Que hoy sea una sola lista es lo que hace ese cambio barato: se
reemplaza este módulo por una consulta y nada más se entera.
"""

AREAS = [
    "TICS",
    "Calidad",
    "SST",
    "Controlados",
    "Facturación",
    "Ventas Institucionales",
    "Mercadeo",
    "Servicio al Cliente",
    "Infraestructura",
    "Logística",
    "Gestión Humana",
    "Contabilidad",
    "Producción",
    "Control Interno",
    "Aseguramiento",
    "Abastecimiento",
    "Comercial",
    "Administración",
    "Tesorería",
    "Puntos de Venta",
]

# Nombres viejos que quedaron en datos ya guardados y a qué área corresponden
# hoy. Las migraciones `d4a8c1f70b32` y `b9e2f4a17c05` los reescribieron en la
# base; esto se queda como documentación de la equivalencia y por si aparece
# un dato rezagado.
EQUIVALENCIAS_HISTORICAS = {
    "TI": "TICS",
    "Sistemas": "TICS",
    "Talento Humano": "Gestión Humana",
    "Gestión humana": "Gestión Humana",
    "Servicio al cliente": "Servicio al Cliente",
    "Ventas institucionales": "Ventas Institucionales",
}


def normalizar(area: str | None) -> str | None:
    """Traduce un nombre viejo al actual. Devuelve None si viene vacío."""
    if not area or not area.strip():
        return None
    limpia = area.strip()
    return EQUIVALENCIAS_HISTORICAS.get(limpia, limpia)


def es_valida(area: str | None) -> bool:
    """None es válido: no todo tiene que tener área asignada."""
    return area is None or area in AREAS

File: protokimica-portal/frontend/src/core/areas.js
/**
 * Las áreas de la empresa. Fuente única del frontend.
 *
 * Antes esta lista estaba repetida en seis archivos y con contenidos
 * distintos entre módulos. Nunca declares una lista de áreas dentro de un
 * componente: impórtala de aquí.
 *
 * El gemelo es `backend/app/core/areas.py`, y una prueba verifica que los dos
 * digan exactamente lo mismo.
 */
export const AREAS = [
  'TICS',
  'Calidad',
  'SST',
  'Controlados',
  'Facturación',
  'Ventas Institucionales',
  'Mercadeo',
  'Servicio al Cliente',
  'Infraestructura',
  'Logística',
  'Gestión Humana',
  'Contabilidad',
  'Producción',
  'Control Interno',
  'Control Interno',
  'Aseguramiento',
  'Abastecimiento',
  'Comercial',

]

/**
 * Nombres viejos que pueden quedar en datos guardados antes de la
 * unificación, y su área actual. Sirve para que un registro histórico no se
 * muestre con un área que ya no existe.
 */
export const EQUIVALENCIAS_HISTORICAS = {
  'TI': 'TICS',
  'Sistemas': 'TICS',
  'Talento Humano': 'Gestión humana',
}

export function normalizarArea(area) {
  if (!area || !area.trim()) return null
  const limpia = area.trim()
  return EQUIVALENCIAS_HISTORICAS[limpia] ?? limpia
}

/**
 * Las áreas a ofrecer en un desplegable, incluyendo el valor actual aunque
 * ya no esté en la lista. Sin esto, editar un registro viejo le borraría el
 * área en silencio al guardar.
 */
export function areasParaSelect(valorActual) {
  if (!valorActual || AREAS.includes(valorActual)) return AREAS
  return [...AREAS, valorActual]
}
